st2common/tests/unit/test_param_utils.py

```diff
@@ -343,34 +343,90 @@ def test_get_finalized_params_with_dict(self):
             'r1': '{{r2}}',
             'r2': {'r2.1': 1},
             'a1': True,
-            'a2': '{{a1}}'
+            'a2': '{{a1}}',
+            'a3': {
+                'test': '{{a1}}',
+                'test1': '{{a4}}',
+                'test2': '{{a5}}',
+            },
+            'a4': 3,
+            'a5': ['1', '{{a1}}']
         }
         runner_param_info = {'r1': {'type': 'object'}, 'r2': {'type': 'object'}}
-        action_param_info = {'a1': {'type': 'boolean'}, 'a2': {'type': 'boolean'}}
+        action_param_info = {
+            'a1': {
+                'type': 'boolean',
+            },
+            'a2': {
+                'type': 'boolean',
+            },
+            'a3': {
+                'type': 'object',
+            },
+            'a4': {
+                'type': 'integer',
+            },
+            'a5': {
+                'type': 'array',
+            },
+        }
         r_runner_params, r_action_params = param_utils.get_finalized_params(
             runner_param_info, action_param_info, params, {})
-        self.assertEqual(r_runner_params, {'r1': {'r2.1': 1}, 'r2': {'r2.1': 1}})
-        self.assertEqual(r_action_params, {'a1': True, 'a2': True})
+        self.assertEqual(
+            r_runner_params, {'r1': {'r2.1': 1}, 'r2': {'r2.1': 1}})
+        self.assertEqual(
+            r_action_params,
+            {
+                'a1': True,
+                'a2': True,
+                'a3': {
+                    'test': True,
+                    'test1': 3,
+                    'test2': [
+                        '1',
+                        True
+                    ],
+                },
+                'a4': 3,
+                'a5': [
+                    '1',
+                    True
+                ],
+            }
+        )
 
     def test_get_finalized_params_with_list(self):
         # Note : In this test runner_params.r1 has a string value. However per runner_param_info the
         # type is an integer. The expected type is considered and cast is performed accordingly.
+        self.maxDiff = None
         params = {
             'r1': '{{r2}}',
             'r2': ['1', '2'],
             'a1': True,
             'a2': 'Test',
             'a3': 'Test2',
             'a4': '{{a1}}',
-            'a5': ['{{a2}}', '{{a3}}']
+            'a5': ['{{a2}}', '{{a3}}'],
+            'a6': [
+                ['{{r2}}', '{{a2}}'],
```

Comment:
Just curious - how many levels deep do we render stuff?

Or is it recursive aka we support arbitrary levels of nesting?